### PR TITLE
Internal improvement: Only run Db code when it is enabled

### PR DESCRIPTION
**Pull request by eggplantzzz**
This PR moves the require statements for @truffle/db into a scope that is run only when db is enabled. I did a few tests and this seems to save an average of 1 second during the compilation flow on my machine when db is not enabled.

There is one possible inefficiency here; in workflow-compile we now require Db 2 times as we are using it in `compileAndSave` and `assignNames`. This might slow the flow down when Db is enabled, or maybe it will be loading it from the cache the second time and there really is no issue here?

Inspired by https://github.com/trufflesuite/truffle/issues/4192

### CODE REVIEW

**Comment by gnidan** [comment]
> This might slow the flow down when Db is enabled, or maybe it will be loading it from the cache the second time and there really is no issue here?

Should be no issue; Node caches modules it requires

**Comment by eggplantzzz** [comment]
Ok so for this we decided to create a new package (db-utils probably) to put the `try/catch` stuff for loading db in and hide it away from our sensitive eyes.

File: packages/core/lib/commands/compile.js
Comment (fainashalts) [inline packages/core/lib/commands/compile.js:94]: ooh! I wonder if the error from #4141 was related to this line!

**Review by fainashalts — COMMENTED** [review]
This LGTM, but I wonder if we should throw an error in db-kit and the db commands (serve and query) to tell the user that they have to enable DB and it must be installed in order to use these? For completeness' sake. What do you think?

File: packages/db-utils/src/index.ts
Comment (cds-amal) [inline packages/db-utils/src/index.ts:1]: What do you think about renaming the function to `getDbOrNull`? 

```suggestion
export const getDbOrNull = () => {
```

Then the use would be clear that the function could return `null` and eliminate the need for a redundant comment (currently this function could return null)

```
let Db = getDbOrNull();
if (Db) {...}
```

File: packages/migrate/Migration.js
Comment (cds-amal) [inline packages/migrate/Migration.js:122]: This comment would not be needed if the function name were changed to `getDbOrNull()` or something similar.

File: packages/core/lib/commands/compile.js
Comment (cds-amal) [reply packages/core/lib/commands/compile.js:94]: @eggplantzzz (sorry, couldn't comment on this without unresolved this conversation)

Does it make sense to move this guard to the `getDb()` function? Of course we'd have to change the interface to accept a `ConfigType` or null, executing the guard only if a config were passed.

**Review by cds-amal — COMMENTED** [review]
Nice work Tyler. I left some comments/questions for a discussion

File: packages/db-utils/src/index.ts
Comment (eggplantzzz) [reply packages/db-utils/src/index.ts:1]: I don't really like `getDbOrNull`, I think it would confuse me if I wasn't familiar. Maybe the comments I wrote are unnecessary.

**Comment by eggplantzzz** [comment]
I also don't think I like the idea of moving the guard. I feel like it makes more sense to put that logic as part of the main command flow and not hidden in the module as my hunch is that it will make what is happening more obvious when reading through the code. Do you disagree?

**Comment by cds-amal** [comment]
I think DB should only be loaded if enabled in config, because it is currently optional and this is an indication the guard belongs in the new module. 

This way would make it more convenient to introduce DB into a new module and revert to a regular import if it stops being an optional dependency (and no longer behind a feature flag)

**Comment by eggplantzzz** [comment]
I'm not really persuaded that it should be moved inside. As of right now I don't think this utility should know more about the config object. It should merely and dumbly load the Db if it can. The instructions for running the command should determine whether db should be loaded in my view. And if it becomes required then it won't really be a hassle at all to just remove the `config.db.enabled`-type checks. Any other opinions on this?

**Comment by fainashalts** [comment]
@eggplantzzz I still think we should use this util folder everywhere DB is being required, which includes `db-kit` and `lib/commands/db`. And in those places, since they are specific-to-db packages, we should throw an error if the `getDb()` call returns null. I feel like this PR isn't complete without adding this functionality.

**Comment by fainashalts** [comment]
Also I think it makes sense to keep the config check in the specific modules, otherwise we have to give the new `db-utils` package access to the config which seems sub-optimal. I don't feel super strongly either way, but that's my two cents!

**Comment by eggplantzzz** [comment]
@fainashalts So I talked with @gnidan and he pointed out that we don't need to use this utility in db-kit because it is a required dependency and will need to be installed to use it. Everywhere else, where it is optional, should use this package.

File: packages/core/lib/commands/db/commands/serve.js
Comment (cds-amal) [inline packages/core/lib/commands/db/commands/serve.js:21]: This shouldn't block the PR, but we should probably guard `db serve` behind the `db: enabled` config/feature flag

**Review by cds-amal — APPROVED** [review]
Nice! Tests pass and looks good to me!

**Review by fainashalts — COMMENTED** [review]
LGTM! Thanks for addressing my previous comments! :)

**Pull request merged**
